Properly propagate more failures and unit test all the things!

petr-muller opened the pull request:
Previously, when we failed to submit some rehearsal jobs, that failure could be covered up later by all successfully submitted rehearsals passing.

Additionally, added a bunch of unit tests for `ExecuteJobs` to do more testing of various, mostly negative, cases.

/cc @droslean @stevekuznetsov @bbguimaraes

petr-muller committed:
Add more unit tests for ExecuteTests

petr-muller committed:
Fail when all jobs attempted to rehearse fail to submit

stevekuznetsov commented on pkg/rehearse/jobs_test.go:
```diff
@@ -254,38 +255,175 @@ func makeTestingProwJob(name, namespace, jobName, context string, refs *pjapi.Re
+				t.Fatalf("Failed to setup watch: %v", err)
+			}
+			fakecs.Fake.PrependWatchReactor("prowjobs", makeSuccessfulFinishReactor(watcher, tc.jobs))
+			fakecs.PrependReactor("create", "prowjobs", tc.reactor)
```
Could explicitly call out to `Fake.PrependReactor` or not above, to be consistent

stevekuznetsov commented on pkg/rehearse/jobs_test.go:
```diff
@@ -254,38 +255,175 @@ func makeTestingProwJob(name, namespace, jobName, context string, refs *pjapi.Re
+	testCases := []struct {
+		description string
+		jobs        map[string][]prowconfig.Presubmit
+		reactor     func(action clientgo_testing.Action) (handled bool, ret runtime.Object, err error)
```
You might be able to DRY this out to a list of job names that could fail on create and make the test cases more narrow

stevekuznetsov commented on pkg/rehearse/jobs_test.go:
```diff
@@ -325,67 +462,58 @@ func TestExecuteJobs(t *testing.T) {
-					createdJobs.Items[i].Status.StartTime.Reset()
-				}
+			// Overwrite dynamic struct members to allow comparison
+			for i := range createdJobs.Items {
```
You could also just validate `[]prowapi.ProwJobSpec`

petr-muller replied:
Yeah, that simplifies everything a bit! :+1:

stevekuznetsov reviewed (commented):
Nice1

petr-muller committed:
Simplify unit tests

petr-muller commented:
@stevekuznetsov Addressed the feedback, thanks. Given some of the simplifications touch tests not related to this PR, I prefer to keep the simplifications in a separate commit, not squash it.

stevekuznetsov commented on pkg/rehearse/jobs_test.go:
```diff
@@ -330,7 +320,17 @@ func TestExecuteJobsErrors(t *testing.T) {
+			fakecs.Fake.PrependReactor("create", "prowjobs", func(action clientgo_testing.Action) (bool, runtime.Object, error) {
+				createAction := action.(clientgo_testing.CreateAction)
+				pj := createAction.GetObject().(*pjapi.ProwJob)
+				for _, jobName := range tc.failToCreate {
```
If you use `sets.String` this becomes `tc.failToCreate.Has(pj.Spec.Job)`

petr-muller replied:
Sweet. It'll take a while until I memorize all the kube apis.

stevekuznetsov reviewed (commented):
Minor nit. otherwise LGTM

droslean commented:
/lgtm
